presto-hive-hadoop2/src/test/java/io/prestosql/plugin/hive/metastore/cache/TestCachingHiveMetastoreWithQueryRunner.java

```diff
@@ -0,0 +1,119 @@
+import java.util.Map;
+import java.util.Optional;
+
+import static com.google.common.io.RecursiveDeleteOption.ALLOW_INSECURE;
+import static io.prestosql.spi.security.SelectedRole.Type.ROLE;
+import static io.prestosql.testing.TestingSession.testSessionBuilder;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+@SuppressWarnings("UnstableApiUsage")   // for Closer
+public class TestCachingHiveMetastoreWithQueryRunner
+{
+    private static final String CATALOG = "test";
+    private static final String SCHEMA = "test";
+    private static final Session ADMIN = getTestSession(new Identity.Builder("admin")
+            .withRole(CATALOG, new SelectedRole(ROLE, Optional.of("admin"))).build());
+    private static final String ALICE_NAME = "alice";
+    private static final Session ALICE = getTestSession(new Identity.Builder(ALICE_NAME).build());
+
+    private DistributedQueryRunner queryRunner;
+    private final Closer closer = Closer.create();
+
+    @BeforeMethod
+    public void createQueryRunner()
+            throws Exception
+    {
+        queryRunner = DistributedQueryRunner
+                .builder(ADMIN)
+                .setNodeCount(1)
+                .build();
+        queryRunner.installPlugin(new HiveHadoop2Plugin());
+        File tmp = Files.createTempDir();
+        closer.register(() -> MoreFiles.deleteRecursively(tmp.toPath(), ALLOW_INSECURE));
+        queryRunner.createCatalog(CATALOG, "hive-hadoop2", Map.of(
+                "hive.metastore", "file",
+                "hive.metastore.catalog.dir", tmp.toURI().toString(),
+                "hive.security", "sql-standard",
+                "hive.metastore-cache-ttl", "60m",
+                "hive.metastore-refresh-interval", "10m"));
+        queryRunner.execute(ADMIN, "CREATE SCHEMA " + SCHEMA);
+        queryRunner.execute("CREATE TABLE test (test INT)");
+    }
+
+    @AfterMethod
+    void cleanUp()
+            throws IOException
+    {
+        closer.close();
+    }
+
+    static Session getTestSession(Identity identity)
+    {
+        return testSessionBuilder()
+                .setCatalog(CATALOG)
+                .setSchema(SCHEMA)
+                .setIdentity(identity)
+                .build();
+    }
+
+    @Test
+    public void testCacheRefreshOnGrantAndRevoke()
+    {
+        assertThatThrownBy(() -> queryRunner.execute(ALICE, "SELECT * FROM test"))
+                .hasMessageContaining("Access Denied");
+        queryRunner.execute("GRANT SELECT ON test TO " + ALICE_NAME);
+        queryRunner.execute(ALICE, "SELECT * FROM test");
+        queryRunner.execute("REVOKE SELECT ON test FROM " + ALICE_NAME);
+        assertThatThrownBy(() -> queryRunner.execute(ALICE, "SELECT * FROM test"))
+                .hasMessageContaining("Access Denied");
+    }
+
+    @DataProvider
+    private Object[][] testCacheRefreshOnRoleGrantAndRevokeParams()
+    {
+        String grantSelectStatement = "GRANT SELECT ON test TO ROLE test_role";
```

Comment:
It is strange that you GRANT something that is created in test (line 112).

Reply:
test table is created in `@BeforeMethod`, so it's quite reasonable

also I wouldn't be surprised if an implementation of "GRANT" would rely not on particular object but on object name, in that case it could even work before creating table